### PR TITLE
[14.0] Add shopfloor_base_multicompany

### PR DESCRIPTION
**Pull request by simahawk**
Provide multi-company support and validation to Shopfloor applications.

TODO: add USAGE/CONFIG instructions.

ref: rau-87

### CODE REVIEW

File: shopfloor_base_multicompany/services/service.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    def _validate_company(self):
+        """
+        Validate current user's companies match one of the companies allowed for the app.
+
+        Raises:
+        exceptions.UserError: If the user's company is not allowed.
+        """
+        if self.collection.must_validate_company:
+            if not self.env.user.company_ids & self.collection.company_ids:
```
Comment (jbaudoux) [inline shopfloor_base_multicompany/services/service.py]: All new records created from the users in the app will likely belong to the company of the user. So we need to ensure a user active in multiple companies is logged in the right company. 
```suggestion
            if not self.env.user.company_id in self.collection.company_ids:
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: shopfloor_base_multicompany/services/service.py (hunk 1 above)
Comment (simahawk) [reply shopfloor_base_multicompany/services/service.py]: fixed

**Review by jbaudoux — APPROVED** [review]
LG

**Comment by simahawk** [comment]
failing test unrelated: `odoo.addons.shopfloor.tests.test_delivery_sublocation: FAIL: DeliveryScanSublocationCase.test_scan_barcode_in_sublocation` 

Should be fixed here https://github.com/OCA/wms/pull/546

File: shopfloor_base_multicompany/tests/test_company_validation.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        )
+        self.shopfloor_app.sudo().company_ids += new_comp
+        self.assertTrue(self.shopfloor_app.must_validate_company)
+
+        service = self.get_service("menu", profile=self.profile)
+        msg = service._validate_company_error_msg()
+        with self.assertRaisesRegex(exceptions.UserError, msg):
+            service.dispatch("search")
+
+        self.env.user.sudo().company_ids += new_comp
```
Comment (jbaudoux) [inline shopfloor_base_multicompany/tests/test_company_validation.py:38]: Needs fixing with last change
```suggestion
        self.env.user.sudo().company_ids += new_comp
        self.env.user.sudo().company_id = new_comp
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
@JuMiSanAr there's a failing test on reception https://github.com/OCA/wms/actions/runs/4075534657/jobs/7022053428#step:8:3276

**Comment by JuMiSanAr** [comment]
> @JuMiSanAr there's a failing test on reception https://github.com/OCA/wms/actions/runs/4075534657/jobs/7022053428#step:8:3276

I've added the fix for this test in #557 .

**Review by jbaudoux — APPROVED** [review]

**Review by jbaudoux — APPROVED** [review]
LG, tested and works

**Pull request merged**